src/nnpack/nnlib.jl

```diff
@@ -0,0 +1,119 @@
+#NOTE: Commenting out the activation functions until sure what to do
+
+# relu(x::AA1) = nnp_relu_output(x, inplace ? x : similar(x), threadpool = shared_threadpool)
+
+# leakyrelu(x::AA1, a = oftype(x/1, 0.01)) =
+#     nnp_relu_output(x, inplace ? x : similar(x), negative_slope = a, threadpool = shared_threadpool)
+
+softmax!(x::A) where A<:AbstractVecOrMat{Float64} = softmax!(Float32.(x))
+
+softmax!(x::A) where A<:AbstractVecOrMat{Float32} =
+    nnp_softmax_output(x, x, threadpool = shared_threadpool)
+
+softmax!(y::A, x::A) where A<:AbstractVecOrMat{Float64} = softmax!(Float32.(y), Float32.(x))
+
+softmax!(y::A, x::A) where A<:AbstractVecOrMat{Float32} =
+    nnp_softmax_output(x, y, threadpool = shared_threadpool)
+
+softmax(x::A) where A<:AbstractVecOrMat{Float64} = softmax(Float32.(x))
+
+softmax(x::A) where A<:AbstractVecOrMat{Float32} =
+    nnp_softmax_output(x, similar(x), threadpool = shared_threadpool)
+
+maxpool(x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float64, 4} =
+    maxpool(Float32.(x), k, pad = pad, stride = stride)
+
+function maxpool(x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float32, 4}
+    pad_, stride_ = expand(Val{length(k)}, pad), expand(Val{length(k)}, stride)
+    ((size(x, 1) - k[1] + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - k[2] + 2 * pad_[2]) % stride_[2] == 0) || error("Choose the stride, pad and kernel size properly")
+    maxpool!(similar(x, pdims(size(x), k, pad_, stride_)), x, k, pad = pad_, stride = stride_)
+end
+
+maxpool!(y::A, x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float64, 4} =
+    maxpool!(Float32.(y), Float32.(x), k, pad = pad, stride = stride)
+
+maxpool!(y::A, x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float32, 4} =
+    nnp_max_pooling_output(x, y, k, padding = expand(Val{length(k)}, pad), stride = expand(Val{length(k)}, stride), threadpool = shared_threadpool)
+
+conv(x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where A<:AbstractArray{Float64, 4} =
+    conv(Float32.(x), Float32.(w), pad = pad, stride = stride, dilation = dilation, algo = algo)
+
+function conv(x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where A<:AbstractArray{Float32, 4}
+    dilation == 1 || dilation == (1, 1) || error("NNPACK does not support dilation > 1")
+    pad_, stride_ = padtuple(x, pad), padtuple(x, stride)
+    ((size(x, 1) - size(w, 1) + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - size(w, 2) + 2 * pad_[2]) % stride_[2] == 0) || error("Choose the stride, pad and kernel size properly")
```

Comment:
Seems like there might be some shared code among these checks, perhaps they can be factored out into a function?